### PR TITLE
Added a more usable string type

### PR DESCRIPTION
**Pull request by jimbxb**
Not yet implemented: non-C-style special resources

### CODE REVIEW

File: wybelibs/wybe/control.wybe [hunk 1]
```diff
@@ -11,13 +11,13 @@ pub def {terminal,semipure} exit(code:int) {
 
 # Print the source location and and error message, and abort the computation.
 
-pub def {terminal,semipure} error(message:string) use call_source_location {
+pub def {terminal,semipure} error(message:c_string) use call_source_location {
```
Comment (pschachte) [inline wybelibs/wybe/control.wybe:14]: There should be a Wybe string version of error.

File: wybelibs/wybe/string.wybe [hunk 2]
```diff
@@ -1,27 +1,145 @@
...
-use wybe.char
+use wybe.int, wybe.c_string, wybe.range
+use wybe.int, wybe.char, wybe.bool
+use wybe.io, wybe.float, wybe.count
+use wybe.comparison
 
-representation is address
+constructor empty
+          | buffer(len:int, raw:c_string) # must be first non-const ctor
+          | concat(left:_, r:_)
```
Comment (pschachte) [inline wybelibs/wybe/string.wybe:10]: Please rename `r` to `right`

**Review by pschachte — CHANGES_REQUESTED** [review]
Looks really good, just a few small changes needed.

File: wybelibs/wybe/control.wybe (hunk 1 above)
Comment (jimbxb) [reply wybelibs/wybe/control.wybe:14]: I agree with this. 

However, I have tried to implement this via the following
```
pub def {terminal,semipure} error(message:string) use call_source_location {
    foreign c {terminal,semipure} error_exit(call_source_location, c_string(message))
}
```
And received this error message: 
```
wybelibs/wybe/control.wybe:14:68: error has normal (total) determinism, but declared terminal
```

It appears there is some underlying issue with determinism checks of terminal procedures. Yet to try to diagnose further

**Review by pschachte — APPROVED** [review]

**Comment by pschachte** [comment]
Uggg!  There are conflicts in four .exp files because of the change I just committed for resources.  Could you please re-run ./update-exp and commit the changes to this PR, so I can merge it?

**Comment by jimbxb** [comment]
On it now. 

Looks like string now needs to import `wybe.phantom`, which wasn't the case before when importing `io`

**Review by pschachte — APPROVED** [review]

**Pull request merged**